### ISSUE TITLE
MemNN example does not reproduce paper results

### ISSUE THREAD

**Issue by nuric** [chainer/chainer#6576]
The [memnn](https://github.com/chainer/chainer/tree/master/examples/memnn) implementation in the examples folder does not seem to reproduce the results reported in the corresponding [paper](https://arxiv.org/abs/1503.08895). In particular, when run on task 2 it doesn't seem to converge to a solution. The implementation was pulled in by #4222.

## Platform
  - Chainer version: 5.3.0
  - CuPy version: running on CPU
  - OS/Platform: Linux 4.15.0-38-generic #41-Ubuntu

### Code to reproduce
We run the instructions from the [readme](https://github.com/chainer/chainer/tree/master/examples/memnn) with task 2 data:
```
python3 train_memnn.py data/tasks_1-20_v1-2/en/qa2_two-supporting-facts_train.txt data/tasks_1-20_v1-2/en/qa
2_two-supporting-facts_test.txt --model qa2model
```
### Error messages, stack traces, or logs
The output of the above command:
```
Training data: data/tasks_1-20_v1-2/en/qa2_two-supporting-facts_train.txt: 200                                        
Test data: data/tasks_1-20_v1-2/en/qa2_two-supporting-facts_test.txt: 200                                             
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy                                 
1           3.50365     3.4641                0.15           0.187                                                    
2           3.43317     3.36324               0.21           0.187                                                    
3           3.30038     3.17261               0.21           0.187                                                    
4           3.05496     2.84275               0.21           0.187                                                    
5           2.66537     2.39886               0.21           0.187                                                    
6           2.25085     2.06654               0.21           0.187                                                    
7           1.99869     1.91604               0.21           0.187
8           1.88459     1.85773               0.21           0.187
9           1.83436     1.84326               0.21           0.187
--------
83          1.15372     2.02805               0.583          0.25
84          1.14651     2.03602               0.592          0.254
85          1.14041     2.05358               0.594          0.255
86          1.13725     2.06104               0.596          0.255
87          1.13028     2.06745               0.602          0.255
88          1.1207      2.07252               0.597          0.257
89          1.11511     2.0785                0.598          0.251
90          1.10864     2.08676               0.602          0.257
91          1.1021      2.10154               0.604          0.254
92          1.09391     2.10309               0.603          0.256
93          1.09071     2.11311               0.605          0.26
94          1.08164     2.12224               0.612          0.255
95          1.07694     2.13073               0.617          0.261
96          1.07041     2.14001               0.617          0.26
97          1.06327     2.14919               0.619          0.266
98          1.05924     2.16001               0.629          0.261
99          1.05256     2.16606               0.629          0.27
100         1.04679     2.17209               0.638          0.261
```

At 100 epochs the model seems to have just overfitted on the task although [paper reference implementation](https://github.com/facebook/MemNN/tree/master/MemN2N-babi-matlab) without linear start converges to a solution.

We run the command 10 times and obtain the same behaviour each time. Is there a test, checked results for the example implementation of the memory network on the bAbI dataset?

**Comment by takagi** [comment]
I've reproduced your result, will look into it further.

**Comment by takagi** [comment]
**1k training examples, BoW**

Task|Test accuracy|Test error rate (%)
-|-|-
1|0.992|0.8
2 | 0.243 | 75.7
3 | 0.224 | 77.6
4 | 0.674 | 32.6
5 | 0.766 | 23.4
6 | 0.482 | 51.8
7 | 0.766 | 23.4
8 | 0.788 | 21.2
9 | 0.599 | 40.1
10 | 0.475 | 52.5
11 | 0.88 | 12
12 | 0.988 | 1.2
13 | 0.923 | 7.7
14 | 0.679 | 32.1
15 | 0.495 | 50.5
16 | 0.465 | 53.5
17 | 0.512 | 48.8
18 | 0.554 | 44.6
19 | 0.09 | 91
20 | 0.916 | 8.4

**Comment by takagi** [comment]
@nuric I've researched the test error rate for each task with 1k training examples and BoW sentence representation on Chainer. I show the result above, following test error rates on the paper.

I can read that task 2, 6, 14 and 20 have low reproducibility but I'm not sure how close the paper's result should be fine. Can you basically agree with the table?

**Comment by nuric** [comment]
@takagi I agree, we are observing a similar trend with our implementation as well. Especially tasks 2, 6 and 14 but if we the run [reference implementation](https://github.com/facebook/MemNN/tree/master/MemN2N-babi-matlab) on task 2 with BoW and no linear start, we get:

Run 1:
```
97 | train error: 0 | val error: 0.20833
98 | train error: 0.0022321 | val error: 0.20833
99 | train error: 0.0022321 | val error: 0.19792
100 | train error: 0.0044643 | val error: 0.20833
test error: 0.18649
```

Run 2:
```
99 | train error: 0.0011161 | val error: 0.14583
100 | train error: 0.0011161 | val error: 0.14583
test error: 0.11089
```

Run 3:
```
98 | train error: 0 | val error: 0.20833
99 | train error: 0.0033482 | val error: 0.20833
100 | train error: 0.0022321 | val error: 0.20833
test error: 0.15827
```

So it seems the reference implementation is producing consistent results with perhaps the extra tricks they employ such as randomising time. Here is the exact configuration of the runs above from `config_babi.m`:

```matlab
batch_size = 32;
nhops = 3;
nepochs = 100;
lrate_decay_step = 25; % reduce learning rate by half every 25 epochs

% use 10% of training data for validation
train_range = 1:floor(0.9 * size(questions,2));
val_range = (floor(0.9 * size(questions,2))+1):size(questions,2);
enable_time = true; % add time embedings
use_bow = true; % use Bag-of-Words instead of Position-Encoding
share_type = 1; % 1: adjecent, 2: layer-wise weight tying 
randomize_time = 0.1; % amount of noise injected into time index
add_proj = false; % add linear layer between internal states
add_nonlin = false; % add non-linearity to internal states

config = {};
config.init_lrate = 0.01;
config.max_grad_norm = 40;
config.input_dim = 20;
config.out_dim = 20;

linear_start = false;
if linear_start
    ls_nepochs = 20;
    ls_lrate_decay_step = 21;
    ls_init_lrate = 0.01/2;
    config.init_lrate = 0.01/2;
end

config.sz = min(50, size(story,2));
config.voc_sz = length(dict);
config.bsz = batch_size;
config.max_words = size(story,1);
if enable_time 
   config.voc_sz = config.voc_sz + config.sz;
   config.max_words = config.max_words + 1; % +1 for time words
end
```

So perhaps we are missing some of those extra configurations in the Chainer implementation that is causing the results to fluctuate (?)

**Comment by takagi** [comment]
Hmm, do you have idea to implement time embeddings and `randomize_time` to Chainer? I'm currently not sure about them and need to understand them first.

**Comment by nuric** [comment]
I thought time embeddings are just implemented by `EmbedID` as usual. The `randomize_time` seems to be a pre-process step that adds empty *sentences* that act as noise so the order of the sentences are not fixed during training. But the important bit is the attention needs to mask those entries out so the model cannot attend to the injected noise memory slots. I'm not sure if this alone would reproduce results either. Perhaps we can say update readme to say it is an approximate implementation of the one from the paper. What do you think?

**Comment by takagi** [comment]
Thanks. I also ask it to the original author of the Chainer version.

@unnonouno The memnn example implementation is supposed to have low reproducibility on some tasks on the paper. Please see the table above for the result I ran the code. Do you have any idea on how much difference of the error rates can be valid on this problem domain, or on what leads the differences against the paper?
